File: tools/parity/dispatch/provider-backoff.ts
// TS dispatch for the provider-backoff parity module: drives the live
// `src/main/rate-limits/active-failure-backoff.ts` reference so the harness
// compares it against the Rust port (orca-provider-backoff).
//
// The Rust core owns base/ceiling as constants (ACTIVE_FAILURE_REFETCH_MS = 30s,
// MAX_ACTIVE_FAILURE_REFETCH_MS = 15min); the TS twin takes them as params. We pin
// the same two values here so the compared pair differs only in language, not in
// configuration — a real behavioural-parity check, not a re-tuned reimplementation.

import { activeFailureRefetchThrottleMs } from '../../../src/main/rate-limits/active-failure-backoff'

const ACTIVE_FAILURE_REFETCH_MS = 30 * 1000
const MAX_ACTIVE_FAILURE_REFETCH_MS = 15 * 60 * 1000

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'activeFailureRefetchThrottleMs': {
      const { streak } = input as { streak: number }
      // Clamp a negative/absent streak to 0, matching the Rust adapter's read of
      // an out-of-range value (both then collapse to the base wait).
      const safeStreak = Number.isFinite(streak) ? Math.max(0, Math.trunc(streak)) : 0
      return activeFailureRefetchThrottleMs(
        safeStreak,
        ACTIVE_FAILURE_REFETCH_MS,
        MAX_ACTIVE_FAILURE_REFETCH_MS
      )
    }
    default:
      return { __parity_error__: `unknown function ${fn}` }
  }
}

File: rust/crates/orca-dispatch/src/modules/provider_backoff.rs
//! Parity dispatch for `orca_provider_backoff` vs the TS twin
//! `src/main/rate-limits/active-failure-backoff.ts`.
//!
//! F4 promotion of an E1-certified, autoformalize-TRUSTED decision core through
//! the one production `orca-dispatch` seam: additive only — no call site is cut
//! over (the shipping throttle stays TS), so this wires the Rust core in for
//! machine-checked parity without a hot-path change. The TS twin parameterizes
//! base/ceiling; the Rust core owns them as constants, so the adapter pins the
//! same `ACTIVE_FAILURE_REFETCH_MS` / `MAX_ACTIVE_FAILURE_REFETCH_MS` values.

use orca_provider_backoff::active_failure_refetch_throttle_ms;
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "activeFailureRefetchThrottleMs" => {
            // A negative or absent streak reads as 0, matching the TS `max(0, …)`
            // guard where a non-positive streak collapses to the base wait.
            let streak = input
                .get("streak")
                .and_then(Value::as_i64)
                .map_or(0, |n| n.max(0).min(u32::MAX as i64) as u32);
            json!(active_failure_refetch_throttle_ms(streak))
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
